quantum/impl/quantum_read_write_spinlock_impl.h

```diff
@@ -27,147 +64,178 @@ namespace quantum {
-        }
-        SpinLockUtil::pauseCPU();
-    }
-    return true;
+    return SpinLockUtil::lockRead(_count, true);
 }
 
 inline
 bool ReadWriteSpinLock::tryLockWrite()
 {
-    int i = 0;
-    return _count.compare_exchange_strong(i, -1, std::memory_order_acq_rel);
+    return SpinLockUtil::lockWrite(_count, true);
 }
 
 inline
-bool ReadWriteSpinLock::unlockRead()
+void ReadWriteSpinLock::unlockRead()
 {
-    int oldValue = 1;
-    int newValue = 0;
-    while (!_count.compare_exchange_weak(oldValue, newValue, std::memory_order_acq_rel))
-    {
-        if (oldValue <= 0)
-        {
-            return false; //unlocked or write-locked
-        }
-        else
-        {
-            newValue = oldValue - 1;
-        }
-        SpinLockUtil::pauseCPU();
-    }
-    return true;
+    SpinLockUtil::unlockRead(_count);
+}
+
+inline
+void ReadWriteSpinLock::unlockWrite()
+{
+    SpinLockUtil::unlockWrite(_count);
 }
 
 inline
-bool ReadWriteSpinLock::unlockWrite()
+void ReadWriteSpinLock::upgradeToWrite()
 {
-    int i = -1;
-    return _count.compare_exchange_strong(i, 0, std::memory_order_acq_rel);
+    SpinLockUtil::upgradeToWrite(_count);
 }
 
 inline
-bool ReadWriteSpinLock::isLocked()
+bool ReadWriteSpinLock::tryUpgradeToWrite()
 {
-    return _count.load(std::memory_order_acquire) != 0;
+    return SpinLockUtil::upgradeToWrite(_count, true);
 }
 
 inline
-bool ReadWriteSpinLock::isReadLocked()
+bool ReadWriteSpinLock::isLocked() const
 {
-    return _count.load(std::memory_order_acquire) > 0;
+    return SpinLockUtil::isLocked(_count);
 }
 
 inline
-bool ReadWriteSpinLock::isWriteLocked()
+bool ReadWriteSpinLock::isReadLocked() const
 {
-    return _count.load(std::memory_order_acquire) == -1;
+    return isLocked() && !isWriteLocked();
+}
+
+inline
+bool ReadWriteSpinLock::isWriteLocked() const
+{
+    return SpinLockUtil::isWriteLocked(_count);
 }
 
 inline
 int ReadWriteSpinLock::numReaders() const
 {
-    int num = _count.load(std::memory_order_acquire);
-    return num == -1 ? 0 : num;
+    return SpinLockUtil::numReaders(_count);
+}
+
+inline
+int ReadWriteSpinLock::numPendingWriters() const
+{
+    return SpinLockUtil::numPendingWriters(_count);
 }
 
 inline
 ReadWriteSpinLock::ReadGuard::ReadGuard(ReadWriteSpinLock& lock) :
     _spinlock(lock),
-    _ownsLock(true)
+    _ownsLock(true),
+    _isUpgraded(false)
 {
     _spinlock.lockRead();
 }
 
 inline
-ReadWriteSpinLock::ReadGuard::ReadGuard(ReadWriteSpinLock& lock, ReadWriteSpinLock::TryToLock) :
+ReadWriteSpinLock::ReadGuard::ReadGuard(ReadWriteSpinLock& lock,
+                                        ReadWriteSpinLock::TryToLock) :
+    _spinlock(lock),
+    _ownsLock(_spinlock.tryLockRead()),
+    _isUpgraded(false)
+{
+}
+
+inline
+ReadWriteSpinLock::ReadGuard::ReadGuard(ReadWriteSpinLock& lock,
+                                        ReadWriteSpinLock::AdoptLock) :
     _spinlock(lock),
-    _ownsLock(_spinlock.tryLockRead())
+    _ownsLock(lock.isReadLocked()),
```

Comment:
Similar to the ReadWriteMutex::ReadGuard(lock, AdoptLock), the logic on _ownsLock here is wrong.